[KIT-64] F/upload

jueleejue opened the pull request:
# Pull Request Check List
- Major Reviewer: 

## CheckList

- [ ] Unit test coverage
- [ ] Backward compatibility
- [ ] Tested on dev/staging environment
- [ ] Documentation
- [ ] Self reviewed

## Context

**PLEASE FILL THIS PAGE**

Commit:
[JPA 연결 확인 위해 controller , service 등 주석처리]
build.gradle 의존성 추가/ Dto 수정/ main에 file upload위한 annotation추가/ upload properties 추가/ VO 추가/ Exception 추가

Commit:
file upload(단일)에 필요한 코드 작성

Commit:
multi upload 추가 / FileUploadService에 service별 디렉토리 존재하지 않으면 생성하는 로직 추가 / 중복파일명 검사 로직 추가

Commit:
새로운 DTO convention에 맞추어 file upload 수정

Commit:
추가된 build.gradle 목록

junhodo commented on build.gradle:
```diff
@@ -21,7 +21,7 @@ dependencies {
 
 	runtimeOnly 'com.h2database:h2'
-	runtimeOnly 'mysql:mysql-connector-java'
+	//runtimeOnly 'mysql:mysql-connector-java'
```
요거는 안쓰는건가요?

junhodo commented:
@juegonjue 
혹시 base가 KIT64인건 의도된건가요?

jueleejue commented:
> @juegonjue
> 혹시 base가 KIT64인건 의도된건가요?

@junhodo 
넹 
main 
⎣ kit-64
⎣⎣     윤진씨(F/download)
⎣⎣    지우(F/upload)
             
이렇게 따로 파서 kit-64에 pr 넣었어요

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+
 
 @Service
+@Transactional(readOnly = true)
```
이거 readOnly로 되어있는데 h2 말고 mysql로 했을때 저장 잘되나요?

jueleejue replied:
확인해보겠습니당

jueleejue replied:
넵 mysql에서 저장 잘 되는 것 확인했습니다 ~!

dldhk97 replied:
지금 보니 클래스에 Transactional Read-only로 되어있는데, 메소드에도 Transactional 어노테이션이 걸려있어서 잘 되는거같네요.

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+  }
 
+  /* 유효성 검증 : 파일명, 파일 크기 */
+  public boolean isValidFile(MultipartFile file) {
```
이 서비스 내에서만 쓰는거라면 외부에서 사용하지 못하도록 private으로 바꾸는게 좋을 것 같습니다.

jueleejue replied:
✅

Bale-do replied:
이 함수는 Valid하지 않으면 error를 throw하고 false는 리턴하지 않네요
boolean 형으로 리턴하고 받는쪽에서 핸들링하거나 혹은 void로 만들어서 에러 발생시 throw만 하는게 좋을 것 같습니다

jueleejue replied:
> 이 함수는 Valid하지 않으면 error를 throw하고 false는 리턴하지 않네요
> boolean 형으로 리턴하고 받는쪽에서 핸들링하거나 혹은 void로 만들어서 에러 발생시 throw만 하는게 좋을 것 같습니다

오옷 감사합니다 수정할게용

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/application/service/exception/MyFileNotFoundException.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.NOT_FOUND)
+public class MyFileNotFoundException extends RuntimeException{
```
기존 NotFoundException이랑 차이가 있으면, 무슨 차이가 있는지 주석을 달아두면 좋을 것 같습니다.
그리고 My는 무엇을 의미하는건가요?

jueleejue replied:
새로운 exception을 반영하지 않았네요. 아마 사용되지 않을 exception같습니다.

jueleejue replied:
✅

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/infra/config/FileUploadProperties.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties(prefix = "se-file-server")
+public class FileUploadProperties {
```
위에서 속성 주입을 어노테이션으로 받고 있으니, 이 클래스를 사용하지 않는다면 지우는게 좋겠습니다.

jueleejue replied:
그러네요 삭제 후 반영하겠습니다.

jueleejue replied:
✅

Bale-do commented on build.gradle:
```diff
@@ -21,7 +21,7 @@ dependencies {
 
 	runtimeOnly 'com.h2database:h2'
-	runtimeOnly 'mysql:mysql-connector-java'
+	//runtimeOnly 'mysql:mysql-connector-java'
```
h2를 사용하려고 주석친게 맞을까요? ci 환경은 mysql로 동작해서 push할떄는 mysql로 올려주시는게 좋을 것 같네용

jueleejue replied:
넹 build.gradle은 제외하고 올릴게요 !
✅

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+  @Value("${se-file-server.upload-dir}")
+  private Path fileLocation;
+
+  /* constructor */
```
요 주석은 필요없을것 같네용

jueleejue replied:
넹 1 ✅

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+  @Transactional
+  public File upload(MultipartFile file, String service) {
+
+    Path targetLocation = this.fileLocation;
```
deep copy인가요 shallow copy인가요?
fileLocation을 건드리는 작업이 있어서 요렇게 진행하는 것 같은데 값이 카피되는데 레퍼런스가 넘어가는지 확인이 필요해 보입니다

jueleejue replied:
완전히 동일한 파일을 새롭게 생성합니다 !

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+    } catch (IOException e) {
+      e.printStackTrace();
+      System.out.println("디렉토리 생성 오류");
+    }
```
service name 별로 디렉터리가 달라서 디렉터리가 없으면 생성하는 로직 같은데 이 로직을 ensureUploadDir()과 같은 함수로 묶어주는건 어떨까요?

jueleejue replied:
✅

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+    String saveName;
+    do {
+      saveName = createSaveName(fileType);
+      System.out.println(saveName);
```
디버깅 코드는 빼는게 좋아보입니당

jueleejue replied:
넹 수정하겠습니다

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+    } while (isSameFileNameExists(targetLocation.resolve(saveName)));
+    String downloadUri = createDownloadUri(saveName);
+
+    File newFile = new File(
```
이 함수에서 "File"이라는 네이밍이 Domain Model의 File, MultipartFile 두 종류인데 모두다 file이라는 네이밍을 써서 가독성이 떨어지는 것 같은데 도메인 모델은 변수명을 fileEntity로 MultipartFile은 multipartFile로 두는건 어떨까요?

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+  }
+
+  @Transactional
+  public File upload(MultipartFile file, String service) {
```
최악의 경우에는 racing condition에 의해 중복 파일명이 생길 수 있는 로직으로 보이네요, 함수 원자성을 보장하는건 어떨까요?

Bale-do commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,134 @@
+  }
+
+  /* 서버에 파일 저장 */
+  private boolean storeFile(MultipartFile file, Path targetLocation) {
```
이 함수도 boolean형이지만 실패의 경우에는 false를 리턴하지 않네요! false를 리턴해 호출측에서 처리하거나 void형으로 만들 수 있을 것 같네요

wonkajin replied:
개인적으로는 void 한표

익셉션에는 모든 트레이스가 담기는 편이 좋은 듯 합니다.

Bale-do commented on src/main/java/com/se/fileserver/v1/file/presentation/FilePresenterFormatter.java:
```diff
@@ -0,0 +1,23 @@
+public class FilePresenterFormatter implements FilePresenter{
+
+  @Override
+  public Response<File> uploadFile(File file) {
```
Domain Model을 직접 참조하는데 dto를 쓰는게 좋을 것 같지만 당장은 이렇게 둬도 괜찮을 것 같네요~

Commit:
파일 예외처리 수정 및 File도메인의 예외 추가.
사용하지 않는 파일 삭제.

wonkajin commented on src/main/java/com/se/fileserver/v1/file/presentation/FilePresenterFormatter.java:
```diff
@@ -0,0 +1,30 @@
+
+  @Override
+  public Response<File> uploadFile(File file) {
+    return new Response<>(HttpStatus.CREATED, "성공적으로 파일을 업로드하였습니다.", file);
```
업로드를 성공하고 굳이 파일 객체 전체를 response 해 줄 필요가 있나요? 파일 경로와 이름이면 충분하지 않을까 싶습니다.

jueleejue replied:
파일경로 = url+이름 으로 구성되어있으니 파일경로 정도만 리턴하는걸로 수정하겠습니당

wonkajin reviewed (commented):
예외를 각 사람마다 다 따로 쓰는 것 같네요. 통일성 있는 예외를 두고 그 처리에 대해서도 작업하시는 분들끼리 논의할 필요가 있어 보입니다.
advice를 쓰고 계신 분도 있구요.

그리고 테스트 코드는 없나요?

Commit:
Service 내에서 원하는 부분만 transaction을 걸어주게 수정.
기존에는 파일 형식 검사 하나씩 하면서 파일업로드 + DB에 저장 하나 되던 로직을 -> 모든 파일들의 형식검사 후 파일 중 하나라도 형식이 어긋나면 올리지 않음.씨 (10개 중 9번째 파일의 형식에 이상이 있으면 8개는 이미 올라가버린 상태라 삭제시키는 비용이 더 클듯함)

Commit:
transaction 재수정.

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,152 @@
+
+    List<File> fileEntityList = new ArrayList<>();
+
+    for (MultipartFile multipartFile : multipartFiles) {
```
반복문 안에 있는 코드는 하나의 파일을 생성하는 코드로 보이는데, 함수로 빼면 가독성이 더 좋아질 것 같습니다.
특히 do ~ while 코드는 UUID로 파일명이 중복되지 않게끔 생성하는 코드인데, 따로 함수로 빼서 적절히 네이밍하면 한눈에 알아볼 수 있을 것 같네요.

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,152 @@
+  }
+
+  /* 유효성 검증 : 파일명, 파일 크기 */
+  private void checkFileCondition(List<MultipartFile> multipartFileList) {
```
함수 네이밍으로는 하나의 파일을 받아서 유효성을 검사하는 것이 더 바람직할 것 같습니다.
반복문을 밖으로 빼고, 이 함수에서는 한 파일을 검사하는 것이 좋을 것 같아요!

dldhk97 commented on src/main/java/com/se/fileserver/v1/file/application/service/FileUploadService.java:
```diff
@@ -1,13 +1,152 @@
+      targetLocation = targetLocation.resolve(directoryName);
+    } catch (IOException e) {
+      e.printStackTrace();
+      System.out.println("디렉토리 생성 오류");
```
시스템 기본 출력보다는 로깅 컨벤션을 따라 slfj로 로그를 남기는 것이 좋을 것 같습니다.

Commit:
file upload service 의 upload 함수 수정.

Commit:
안쓰는 함수 삭제

Commit:
add : 파일 목록 조회

Commit:
Merge remote-tracking branch 'origin/main' into F/upload

# Conflicts:
#	src/main/java/com/se/fileserver/v1/FileserverApplication.java
#	src/main/java/com/se/fileserver/v1/common/application/dto/ExceptionResponse.java
#	src/main/java/com/se/fileserver/v1/common/application/dto/request/BaseRequest.java
#	src/main/java/com/se/fileserver/v1/common/application/dto/request/PaginationRequest.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/AttachmentTooLargeException.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/BadRequestException.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/InvalidInputException.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/PermissionDeniedException.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/PreconditionFailedException.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/UnauthenticatedException.java
#	src/main/java/com/se/fileserver/v1/common/domain/exception/UniqueValueAlreadyExistsException.java
#	src/main/java/com/se/fileserver/v1/common/infra/springboot/advice/SeExceptionAdvice.java
#	src/main/java/com/se/fileserver/v1/common/presentation/response/PaginationResponse.java
#	src/main/java/com/se/fileserver/v1/common/presentation/response/Response.java
#	src/main/java/com/se/fileserver/v1/file/application/service/FileDeleteService.java
#	src/main/java/com/se/fileserver/v1/file/application/service/FileDownloadService.java
#	src/main/java/com/se/fileserver/v1/file/domain/model/File.java
#	src/main/java/com/se/fileserver/v1/file/domain/repository/FileRepository.java
#	src/main/java/com/se/fileserver/v1/file/domain/repository/FileRepositoryProtocol.java
#	src/main/java/com/se/fileserver/v1/file/infra/repository/FileJpaRepository.java

Commit:
fix

Commit:
불필요한 파일 삭제

Commit:
fix : 파일 업로드 및 파일 목록 읽기 + build.gradle추가

Commit:
add : 파일(단일)업로드 (FE편의성)

Commit:
fix : file type

Commit:
fix : saveName unique in DB & directory

Commit:
fix : upload & download uri

dldhk97 committed:
Remove FileTest.java

dldhk97 committed:
Remove FileDownloadServiceTest.java

dldhk97 reviewed (approved):
LGTM
